File: ea-robot-react-native/modules/device-telemetry/android/src/main/java/com/earobot/devicetelemetry/DeviceTelemetryModule.kt
package com.earobot.devicetelemetry

import android.os.Build
import android.os.PowerManager
import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DeviceTelemetryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DeviceTelemetry")

    Events("onThermalStateChange")

    AsyncFunction("getThermalStateAsync") {
      val powerManager = appContext.reactContext?.getSystemService(Context.POWER_SERVICE) as? PowerManager
      val status = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        powerManager?.currentThermalStatus ?: 0
      } else {
        0
      }
      getThermalStatusString(status)
    }

    OnStartObserving {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        val powerManager = appContext.reactContext?.getSystemService(Context.POWER_SERVICE) as? PowerManager
        powerManager?.addThermalStatusListener { status ->
          sendEvent("onThermalStateChange", mapOf(
            "state" to getThermalStatusString(status)
          ))
        }
      }
    }
  }

  private fun getThermalStatusString(status: Int): String {
    return when (status) {
      PowerManager.THERMAL_STATUS_NONE -> "NOMINAL"
      PowerManager.THERMAL_STATUS_LIGHT -> "FAIR"
      PowerManager.THERMAL_STATUS_MODERATE -> "SERIOUS"
      PowerManager.THERMAL_STATUS_SEVERE -> "CRITICAL"
      PowerManager.THERMAL_STATUS_CRITICAL -> "CRITICAL"
      PowerManager.THERMAL_STATUS_EMERGENCY -> "CRITICAL"
      PowerManager.THERMAL_STATUS_SHUTDOWN -> "CRITICAL"
      else -> "UNKNOWN"
    }
  }
}

File: ea-robot-react-native/modules/device-telemetry/ios/DeviceTelemetryModule.swift
import ExpoModulesCore

public class DeviceTelemetryModule: Module {
  public override func definition() -> ModuleDefinition {
    Name("DeviceTelemetry")

    Events("onThermalStateChange")

    AsyncFunction("getThermalStateAsync") { () -> String in
      return getThermalStateString(ProcessInfo.processInfo.thermalState)
    }

    OnStartObserving {
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(self.thermalStateChanged),
        name: ProcessInfo.thermalStateDidChangeNotification,
        object: nil
      )
    }

    OnStopObserving {
      NotificationCenter.default.removeObserver(self)
    }
  }

  @objc
  private func thermalStateChanged() {
    let state = ProcessInfo.processInfo.thermalState
    sendEvent("onThermalStateChange", [
      "state": getThermalStateString(state)
    ])
  }

  private func getThermalStateString(_ state: ProcessInfo.ThermalState) -> String {
    switch state {
    case .nominal:
      return "NOMINAL"
    case .fair:
      return "FAIR"
    case .serious:
      return "SERIOUS"
    case .critical:
      return "CRITICAL"
    @unknown default:
      return "UNKNOWN"
    }
  }
}
